File: stream-translator/src/main/java/es/upm/fi/dia/oeg/integration/translation/QueryOptimizer.java
package es.upm.fi.dia.oeg.integration.translation;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.Map.Entry;

import org.apache.commons.lang.NotImplementedException;
import org.apache.log4j.Logger;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.hp.hpl.jena.sparql.algebra.Op;

import es.upm.fi.dia.oeg.integration.algebra.OpBinary;
import es.upm.fi.dia.oeg.integration.algebra.OpInterface;
import es.upm.fi.dia.oeg.integration.algebra.OpJoin;
import es.upm.fi.dia.oeg.integration.algebra.OpMultiUnion;
import es.upm.fi.dia.oeg.integration.algebra.OpProjection;
import es.upm.fi.dia.oeg.integration.algebra.OpRelation;
import es.upm.fi.dia.oeg.integration.algebra.OpSelection;
import es.upm.fi.dia.oeg.integration.algebra.OpUnary;
import es.upm.fi.dia.oeg.integration.algebra.OpUnion;
import es.upm.fi.dia.oeg.integration.algebra.xpr.BinaryXpr;
import es.upm.fi.dia.oeg.integration.algebra.xpr.OperationXpr;
import es.upm.fi.dia.oeg.integration.algebra.xpr.ValueXpr;
import es.upm.fi.dia.oeg.integration.algebra.xpr.VarXpr;
import es.upm.fi.dia.oeg.integration.algebra.xpr.Xpr;

public class QueryOptimizer
{
	private static Logger logger = Logger.getLogger(QueryOptimizer.class.getName());
	
	public OpInterface staticOptimize(OpBinary op)
	{
		OpInterface l = staticOptimize(op.getLeft());
		OpInterface r = staticOptimize(op.getRight());
		op.setLeft(l);
		op.setRight(r);
		
		if (op instanceof OpJoin)
		{
			OpJoin join = (OpJoin)op;
			if (l == null || r == null)
			{
				return null;  //join with empty relations is empty
			}
			else if (join.getLeft() instanceof OpProjection &&
				join.getRight() instanceof OpProjection)
			{
				/*
				OpProjection pl = (OpProjection)join.getLeft();
				OpProjection pr = (OpProjection)join.getRight();
				if (pl.getSubOp() instanceof OpRelation && pr.getSubOp() instanceof OpRelation)
				{
					OpRelation rl = (OpRelation)pl.getSubOp();
					OpRelation rr = (OpRelation)pr.getSubOp();
					if (join.hasEqualConditions())//(rl.getExtentName().equals(rr.getExtentName()))
					{
						logger.debug("Merge inner projections of join: "+pl+"-"+pr);
						for (String key : pr.getExpressions().keySet())//merge moth sides
						{
							if (pl.getExpressions().containsKey(key) && 
									pl.getExpressions().get(key) instanceof VarXpr)
							{
								VarXpr var = (VarXpr)pl.getExpressions().get(key);
								if (var.getModifier()!=null)
									continue;
							}
							pl.addExpression(key, pr.getExpressions().get(key));
						}
						return pl;
					}
					else if (rr.getExtentName().equals("constants")) //TODO test for left side
					{
						for (String key : pr.getExpressions().keySet())
						{
							pl.addExpression(key, pr.getExpressions().get(key));
						}
						return pl;
					}
					else if (!join.getConditions().isEmpty())
					{
						//throw new NotImplementedException("join with two dispaired projs");
					}
				}*/
			}
			else
			{
				logger.debug("join merging "+join.getConditions());
				join.getLeft().display();
				join.getRight().display();
				if (join.getConditions().isEmpty())
				{
					logger.debug("Cross product");
					return join;
				}

				if (join.getLeft() instanceof OpJoin)
				{
					OpJoin ljoin = (OpJoin)join.getLeft();
					if (ljoin.getConditions().isEmpty())// optimize empty cross prod
					{
						join.setLeft(ljoin.getRight());
						ljoin.setRight(join);
						return ljoin;
					}
				}
					
				if (join.hasEqualConditions())
				{
					OpInterface merged = join.getLeft().merge(join.getRight(),join.getConditions());
					//if (merged !=null)
					logger.debug("merged join");
						merged.display();
					return merged;
				}
				
				
				if (join.getLeft() instanceof OpMultiUnion)
					//join.getRight() instanceof OpMultiUnion) //push down join in unions
				{
					OpMultiUnion newunion = new OpMultiUnion(null);
					
					OpMultiUnion lunion = (OpMultiUnion)join.getLeft();
					
					
					Collection<OpInterface> rightOps = null;
					if (join.getRight() instanceof OpMultiUnion)
					{
						OpMultiUnion runion = (OpMultiUnion)join.getRight();
						rightOps = runion.getChildren().values();
					}
					else if (join.getRight() instanceof OpProjection)
					{
						rightOps = Lists.newArrayList();
						rightOps.add(join.getRight());
					}
					
					
					int i=0;
					for (OpInterface lo:lunion.getChildren().values())
					{
						for (OpInterface ro:rightOps)
						{
							OpJoin newjoin = new OpJoin(null,lo.copyOp(),ro.copyOp(), false );
							newjoin.getConditions().addAll(join.getConditions());
							newunion.getChildren().put("join"+i, newjoin);
							i++;
						}
					}
							
					return newunion;
				}
					
				/*
				else if (join.getRight() instanceof OpProjection)					
				{
					OpProjection proj = (OpProjection)join.getRight();
					if (proj.getRelation().getExtentName().equals("constants"))
					{
						OpInterface merged = join.getLeft().merge(join.getRight());				;
						return merged;
					}
				}*/
			}
		}
		else if (op.getName().equals("union"))
		{
			//if (r==null)
				//logger.info("fuckkkkkk");
		}
		/*
		if (l == null && r!= null)
			return r;
		else if (r==null && l != null)
			return l;
		else if (r == null & l == null) 
			return null;		
		else return op;*/
		return op;
	}
	public OpInterface staticOptimize(OpMultiUnion union)
	{
		logger.debug("optimize: "+union);
		//union.display();
		Set<String> toRemove = Sets.newHashSet();
		for (Entry<String,OpInterface> entry: union.getChildren().entrySet())
		{	
			OpInterface optimized = staticOptimize(entry.getValue());
			if (optimized!=null)
				union.getChildren().put(entry.getKey(),optimized );
			else
			{
				toRemove.add(entry.getKey());				
			}
		}
		for (String key:toRemove)
		{
			union.getChildren().remove(key);
			for (String k:union.index.keySet())
			{
				if (union.index.get(k).containsValue(key))
				{
					String rkey = null;
					for (String k2:union.index.get(k).keySet())
					{
						if (union.index.get(k).get(k2).contains(key))
							rkey = k2;
					}
					union.index.get(k).removeAll(rkey);
				}
				//throw new NotImplementedException("this key "+key);
				//union.index.get(k).inverse().remove(key);
			}
		}
		if (union.getChildren().size()==1)
		{
			logger.debug("simplified");
			//return union.getChildren().values().iterator().next();
		}
		//union.display();
		return union;
		
	}
	
	public OpInterface staticOptimize(OpUnary op)
	{
		OpInterface o =staticOptimize(op.getSubOp());
		op.setSubOp(o);
		
		if (op.getName().equals("projection"))
		{
			
			OpProjection proj = (OpProjection)op;
			if (proj.getSubOp()==null) //no relation to project
			{
				return null;
			}
			else if (proj.getSubOp().getName().equals("selection"))
			{
				OpSelection sel = (OpSelection)proj.getSubOp();
				logger.debug("remove constant selections");
				for (Xpr xpr:sel.getExpressions())
				{
				BinaryXpr cond = (BinaryXpr) xpr;
				if (cond.getOp().equals("=") && cond.getLeft() instanceof VarXpr)
				{
					VarXpr var = (VarXpr)cond.getLeft();  //so ugly code, we can refactor this mess
					if (proj.getExpressions().containsKey(var.getVarName()) && 
							proj.getExpressions().get(var.getVarName()) instanceof OperationXpr )
					{
						OperationXpr oper = (OperationXpr)proj.getExpressions().get(var.getVarName());
						if (oper.getOp().equals("constant"));
						{
							ValueXpr val = (ValueXpr)cond.getRight();
							if (val.getValue().equals(((ValueXpr) oper.getParam()).getValue()))
							{								
								proj.setSubOp(sel.getSubOp());
								return proj;
							}
							else
							{
								return null;
							}
						}
					}
				}
			
				}
				
			}
			else if (proj.getSubOp() instanceof OpBinary ) //push down projection in unions :)
			{
				OpBinary union = (OpBinary)proj.getSubOp();
				
				
				
				OpProjection leftProj = proj.copyOp();
				OpProjection rightProj = proj.copyOp();
				if (proj.getSubOp() instanceof OpJoin)
				{
					OpJoin join = (OpJoin)proj.getSubOp();
					for (Xpr xpr:join.getConditions())
					{
						for (String var:xpr.getVars())
						{
							if (!proj.getExpressions().containsKey(var))
							{
								leftProj.addExpression(var,new VarXpr(var));
								rightProj.addExpression(var,new VarXpr(var));
							}
						}
					}
				}
				leftProj.setSubOp(union.getLeft());
				rightProj.setSubOp(union.getRight());
				OpInterface l = staticOptimize(leftProj);
				OpInterface r = staticOptimize(rightProj);
				union.setLeft(l);
				union.setRight(r);
				//should simplify here!!!!
				return	union.simplify();
			
				//return union;
			}
			else if (o instanceof OpMultiUnion)
			{
				OpMultiUnion union = (OpMultiUnion)o;
				Map<String,OpInterface> newChildren = new TreeMap<String, OpInterface>();
				for (Map.Entry<String, OpInterface> col:union.getChildren().entrySet())
				{
					OpProjection copy= new OpProjection(proj.getId(),null);
					copy.getExpressions().putAll(proj.getExpressions());
					//Collection<OpInterface> newCol = new ArrayList<OpInterface>();
					//for (OpInterface opi:col.getValue())
					
						copy.setSubOp(col.getValue());
						OpInterface newOp = ( staticOptimize(copy));
					
					newChildren.put(col.getKey(),newOp);

				}
				union.setChildren(newChildren);
				return union;
			}
			else
			{
				if (o instanceof OpProjection ) //double projection, eat the inner!!!
				{
					logger.debug("merge double projections");
					OpProjection innerProj = (OpProjection)o;
					//boolean test = true;
					for (String key: proj.getExpressions().keySet())
					{
						if (innerProj.getExpressions().containsKey(key))
						{	
							proj.getExpressions().put(key, innerProj.getExpressions().get(key));
							//test=false;
							//break;
						}
						else //inner projection doesn't cover the attributes of the outer one
						{
							proj.getExpressions().put(key, ValueXpr.NullValueXpr);
						}
					}
					//if (test)
						proj.setSubOp(innerProj.getSubOp());
					//else
						//proj.setSubOp(o);
				}
				else if (o instanceof OpUnion)
				{
					System.out.println("cosos");
					/*
					OpUnion union = (OpUnion)o;
					OpInterface l = staticOptimize(union.getLeft());
					OpInterface r = staticOptimize(union.getRight());
					union.setRight(r);
					union.setLeft(l);
					return union;*/
				}
				else if (o!=null)  
					proj.setSubOp(o);
				else   // if projection has no relation to project!!!
					return null;
				
				return proj;
			}
		}
		else if (op instanceof OpSelection)
		{
			OpSelection selection = (OpSelection)op;
			//BinaryXpr bin = (BinaryXpr) xpr;
			//VarXpr var = (VarXpr) bin.getLeft();
			if (op.getSubOp() instanceof OpProjection) //apply the selection to the projection
			{
				OpProjection proj = (OpProjection)op.getSubOp();
				for (String varName:selection.getSelectionVars())
				{
					if (proj.getExpressions().containsKey(varName))
					{
						OpSelection sel = new OpSelection(selection.getId(),null,selection.getExpressions());
						proj.build(sel);
						return proj;
					}
					else
						return null;
				}
			}
			else if (op.getSubOp() instanceof OpUnion)
			{
				// push down selection to union children
				OpUnion union = (OpUnion)op.getSubOp();
				OpSelection selLeft = new OpSelection(selection.getId(),union.getLeft(),selection.getExpressions());
				//selLeft.getExpressions().addAll(selection.getExpressions());
				OpSelection selRight = new OpSelection(selection.getId(),union.getRight(),selection.getExpressions());
				//selRight.getExpressions().addAll(selection.getExpressions());
				OpInterface left = staticOptimize(selLeft);
				OpInterface right = staticOptimize(selRight);
				return new OpUnion(left, right).simplify();
				
			}
			else if (op.getSubOp() instanceof OpMultiUnion)
			{
				OpMultiUnion union = (OpMultiUnion)op.getSubOp();
				TreeMap<String,OpInterface> newChildren = new TreeMap<String, OpInterface>();
				for (Map.Entry<String, OpInterface> col:union.getChildren().entrySet())
				{
					//Collection<OpInterface> col2 = new ArrayList<OpInterface>();
					//for (OpInterface opi:col.getValue())
					
					OpSelection sel = new OpSelection(selection.getId(),col.getValue(),selection.getExpressions());
					//sel.getExpressions().addAll(selection.getExpressions());
					OpInterface newChild = staticOptimize(sel);
					//col2.add(newChild);
					
					newChildren.put(col.getKey(), newChild);
				}
				union.setChildren(newChildren);
				return union;
			}
			else if (op.getSubOp() instanceof OpJoin)
			{
				OpJoin join = (OpJoin)op.getSubOp();
				join.getConditions().addAll(selection.getExpressions());
				return join;
				//throw new NotImplementedException("Not done for joins");
			}
			
		}
		else
		{
			OpInterface oi =staticOptimize(op.getSubOp());
			op.setSubOp(oi);
			return op;
		}
		return op;
	}
	
	public OpInterface staticOptimize(OpInterface op)
	{
		if (op instanceof OpUnary)
			return staticOptimize((OpUnary)op);
		else if (op instanceof OpBinary)
			return staticOptimize((OpBinary)op);
		else if (op instanceof OpMultiUnion)
			return staticOptimize((OpMultiUnion)op);
		
		
		
		return op;
	}
}

File: query-rewriting/src/main/scala/es/upm/fi/oeg/morph/stream/rewriting/QueryOptimizer.scala
package es.upm.fi.oeg.morph.stream.rewriting
import collection.JavaConversions._
import java.util.TreeMap
import org.apache.commons.lang.NotImplementedException
import es.upm.fi.oeg.morph.stream.algebra.LeftOuterJoinOp
import es.upm.fi.oeg.morph.stream.algebra.InnerJoinOp
import es.upm.fi.oeg.morph.stream.algebra.MultiUnionOp
import es.upm.fi.oeg.morph.stream.algebra.ProjectionOp
import es.upm.fi.oeg.morph.stream.algebra.AlgebraOp
import es.upm.fi.oeg.morph.stream.algebra.BinaryOp
import es.upm.fi.oeg.morph.stream.algebra.UnaryOp
import es.upm.fi.oeg.morph.stream.algebra.SelectionOp
import es.upm.fi.oeg.morph.stream.algebra.xpr.BinaryXpr
import es.upm.fi.oeg.morph.stream.algebra.xpr.VarXpr
import es.upm.fi.oeg.morph.stream.algebra.xpr.OperationXpr
import es.upm.fi.oeg.morph.stream.algebra.xpr.ValueXpr
import es.upm.fi.oeg.morph.stream.algebra.xpr.NullValueXpr
import es.upm.fi.oeg.morph.stream.algebra.LeftOuterJoinOp
import es.upm.fi.oeg.morph.stream.algebra.RootOp
import es.upm.fi.oeg.morph.stream.algebra.ProjectionOp
import es.upm.fi.oeg.morph.stream.algebra.JoinOp
import es.upm.fi.oeg.morph.stream.algebra.xpr.XprUtils
import es.upm.fi.oeg.morph.stream.algebra.RelationOp
import es.upm.fi.oeg.morph.stream.algebra.GroupOp
import es.upm.fi.oeg.morph.stream.algebra.xpr.AggXpr
import org.slf4j.LoggerFactory
import es.upm.fi.oeg.morph.stream.algebra.xpr.Xpr
import scala.compat.Platform
import scala.util.Random

class QueryOptimizer{
  private val logger= LoggerFactory.getLogger(this.getClass)

  def staticOptimize(op:AlgebraOp):AlgebraOp={
	if (op==null) op
	else op match {
	  case unary:UnaryOp=>staticOptimize(unary)
	  case binary:BinaryOp=>staticOptimize(binary)
	  case union:MultiUnionOp=>staticOptimize(union)
	  case _ =>op
	}
  }
  
  def staticOptimize(op:UnaryOp):AlgebraOp= {
	val o = staticOptimize(op.subOp)
	op match {
	  case proj:ProjectionOp=>	
		optimizeProjection(new ProjectionOp(proj.expressions,o,proj.distinct))
	  case selection:SelectionOp=>
	    optimizeSelection(new SelectionOp(selection.id,o,selection.expressions))
	  case root:RootOp=> new RootOp(op.id,o)
	  case group:GroupOp=>
        optimizeGroup(new GroupOp(group.id,group.groupVars,group.aggs,o))
	  case _ => op
	}
  }
  
  def staticOptimize(op:BinaryOp):AlgebraOp={
	val l = staticOptimize(op.left)
	val r = staticOptimize(op.right)	
	op match {
	  case join:InnerJoinOp=>optimizeJoin(new InnerJoinOp(l,r))
	  case join:LeftOuterJoinOp=>optimizeLeftJoin(new LeftOuterJoinOp(l,r))
	  case _=>op
	}
  }
  
  def staticOptimize(union:MultiUnionOp):AlgebraOp={
	logger.trace("optimize union: "+union.toString)		
	if (logger.isTraceEnabled)
	  union.display
	val newChildren=union.children.values.map(op=>staticOptimize(op)).filter(_!=null).zipWithIndex
	val newUnion=new MultiUnionOp(newChildren.map(c=>c._1.id+c._2->c._1).toMap)
	val u=newUnion.simplify
	if (logger.isTraceEnabled){
	  logger.trace("after union optim")
	  u.display
	}
	u
  }
	
  def optimizeLeftJoin(join:LeftOuterJoinOp):AlgebraOp={
    if (join.left == null || join.right == null)
	  null  //join with empty relations is empty
	else if (join.conditions.isEmpty){
	  logger.debug("Cross product join")
	  join
	}
	else {
	  val pushed=pushDownLeftJoin(join)
	  if (pushed==join) join
	  else staticOptimize(pushed) 	    
	}    
  }

  def pushDownJoin(join:InnerJoinOp):AlgebraOp={
    (join.left,join.right) match{
      case (op1:ProjectionOp,op2:ProjectionOp)=>
        logger.trace("push down join "+ join.conditions.mkString("--")+" ")
        logger.trace(join.hasEqualConditions+ join.isCompatible.toString+op1.getRelation)
        if (join.hasEqualConditions && join.isCompatible && 
            op1.getRelation.extentName.equals(op2.getRelation.extentName) &&
            //op1.subOp.isInstanceOf[RelationOp] && op2.subOp.isInstanceOf[RelationOp] &&
            join.isJoinOnPk  ){
          logger.debug("merging projs")
          op1.merge(op2)
        }
        else if (!join.isCompatible) null
        else join        
        
      case (op1:MultiUnionOp,op2:AlgebraOp)=>
        MultiUnionOp(op1.children.map(c=>c._1->optimizeJoin(new InnerJoinOp(c._2,op2))))
      case (op1:AlgebraOp,op2:MultiUnionOp)=>
        MultiUnionOp(op2.children.map(c=>c._1->optimizeJoin(new InnerJoinOp(c._2,op1))))

      case (op1:AlgebraOp,op2:JoinOp)=>
        
        val optim=optimizeJoin(new InnerJoinOp(op1,op2.left))
        val optim2=optimizeJoin(new InnerJoinOp(op1,op2.right))
        if (!optim.isInstanceOf[JoinOp]) optimizeJoin(new InnerJoinOp(optim,op2.right))
        else if (!optim2.isInstanceOf[JoinOp]) optimizeJoin(new InnerJoinOp(optim2,op2.left))
        else join
      case (op1:JoinOp,op2:AlgebraOp)=>
        val optim=optimizeJoin(new InnerJoinOp(op2,op1.left))
        val optim2=optimizeJoin(new InnerJoinOp(op2,op1.right))
        if (!optim.isInstanceOf[JoinOp]) optimizeJoin(new InnerJoinOp(optim,op1.right))
        else if (!optim2.isInstanceOf[JoinOp]) optimizeJoin(new InnerJoinOp(optim2,op1.left))
        else join
      case _=>join
    }
  }
  
  def pushDownLeftJoin(join:LeftOuterJoinOp):AlgebraOp={
    (join.left,join.right) match{
      case (op1:ProjectionOp,op2:ProjectionOp)=>
        logger.debug("push down join "+ join.conditions.mkString("--")+" ")
        logger.debug(join.hasEqualConditions+ join.isCompatible.toString+op1.getRelation)
        if (join.hasEqualConditions && join.isCompatible && 
            op1.getRelation.extentName.equals(op2.getRelation.extentName) &&
            op1.subOp.isInstanceOf[RelationOp] && op2.subOp.isInstanceOf[RelationOp] &&
            join.isJoinOnPk  ){
          logger.debug("merging projs")
          op1.merge(op2)
        }
        else if (!join.isCompatible) null
        else join        
        
      case (op1:MultiUnionOp,op2:AlgebraOp)=>
        MultiUnionOp(op1.children.map(c=>c._1->optimizeJoin(new InnerJoinOp(c._2,op2))))
      case (op1:AlgebraOp,op2:MultiUnionOp)=>
        MultiUnionOp(op2.children.map(c=>c._1->optimizeJoin(new InnerJoinOp(c._2,op1))))
      case _=>join
    }
  }
  
  /*
  def listCombinations(join:JoinOp)=(join.left,join.right) match{
    case (op1:ProjectionOp,op2:ProjectionOp)=>List(join)
    case (op1:MultiUnionOp,op2:AlgebraOp)=>
    
  }*/
  
  def optimizeJoin(join:InnerJoinOp):AlgebraOp={
    //join.display
    if (join.left == null || join.right == null)
	  return null  //join with empty relations is empty
	if (join.conditions.isEmpty && false){
	  logger.debug("Cross product")
	  return join
	}		
	else {
	  //join.display
	  val pushed=pushDownJoin(join)
	  if (pushed==join) return join
	  //else if (pushed.isInstanceOf[JoinOp]) return pushed
	  else return staticOptimize(pushed) 
	}
    //return join
  }
  

  def optimizeProjection(proj:ProjectionOp):AlgebraOp={
    val o =proj.subOp
    if (proj.subOp==null) //no relation to project
	  null
	else o match {
	  case bin:BinaryOp=>
		val addVars=if (proj.subOp.isInstanceOf[InnerJoinOp]){
		  val join=proj.subOp.asInstanceOf[InnerJoinOp]
		  join.conditions.map{xpr=>xpr.varNames.filter(vr=> !proj.expressions.contains(vr))
					}.flatten.map(name=>name->VarXpr(name))
		  }
		  else List()
				
		val l = staticOptimize(bin.left);
		val r = staticOptimize(bin.right);				
		val newBin=proj.subOp match {
		  case inner:InnerJoinOp=> new InnerJoinOp(l,r)
		  case outer:LeftOuterJoinOp=> new LeftOuterJoinOp(l,r)
		  case _=>throw new Exception("unsupported optimization")
		}
		//should simplify here!!!!
		new ProjectionOp(proj.expressions,newBin,proj.distinct)
	  case union:MultiUnionOp=>	
	    println("befooooore ")
	    
	    union.display
		val newChildren = 
		  union.children.entrySet.map{col=>
		    println(col.getKey())
		    col.getKey+Random.nextDouble->staticOptimize(new ProjectionOp(proj.expressions, col.getValue,proj.distinct))
		  }.toMap
		  println(newChildren.size)
		val mu=new MultiUnionOp(newChildren)//.simplify
		println("pibibi")
		mu.display
		mu
		//throw new Exception("got here")
	  case group:GroupOp=>
		val newExpr=proj.expressions.map{e=>
		  if (group.aggs.contains(e._1)) 
			(e._1,VarXpr(e._1))
		  else e
		}
		new ProjectionOp(newExpr,o,proj.distinct)
	  case innerProj:ProjectionOp =>//double projection, eat the inner!!!		
		logger.debug("merge double projections "+proj+"\n"+o);
		
		val xprs=
			proj.expressions.keySet.map{key=>
			  if (innerProj.expressions.containsKey(key))
				(key, innerProj.expressions(key))
			  else //inner projection doesn't cover the attributes of the outer one
			    (key, NullValueXpr)
			}.toMap
		new ProjectionOp(proj.expressions++xprs,innerProj.subOp,proj.distinct)				
	  case _=>			
		if (o!=null)  
		  new ProjectionOp(proj.expressions,o,proj.distinct)
		else  null // if projection has no relation to project!!!
		  				
	}		
		
  }
	
  private def replaceXpr(xpr:Xpr,xprs:Map[String,Xpr]):Xpr=xpr match{
    case bin:BinaryXpr=>
      new BinaryXpr(bin.op,replaceXpr(bin.left,xprs),replaceXpr(bin.right,xprs))      
    case varXpr:VarXpr=>
      xprs.getOrElse(varXpr.varName, varXpr)    
    case _=>xpr
  }
  
  def optimizeSelection(selection:SelectionOp):AlgebraOp=selection.subOp match {
    case proj:ProjectionOp=>
        val found=selection.selectionVarNames.map(varName=>
		  proj.expressions.containsKey(varName))
		if (found.forall(_==true)){	
		    val newExprs=selection.expressions.map(x=>replaceXpr(x,proj.expressions))
		    val sel = new SelectionOp(selection.id,proj.subOp,newExprs)
			new ProjectionOp(proj.expressions,sel,proj.distinct)
		}
		else proj
	case union:MultiUnionOp=>
		val newChildren = new TreeMap[String, AlgebraOp]
		union.children.entrySet.foreach{col=>					
		    val sel = new SelectionOp(selection.id,col.getValue(),selection.expressions);
			val newChild = staticOptimize(sel);
			newChildren.put(col.getKey(), newChild);
		}
		new MultiUnionOp(union.children++ newChildren).simplify	
	case join:InnerJoinOp=>
		logger.debug("selection inside join "+join.toString)
		//val selxprs=selection.expressions.map(_.asInstanceOf[BinaryXpr]).toList
		val selVars=selection.expressions.map(x=>x.varNames).flatten
		if (join.left.vars.containsAll(selVars) && join.right.vars.containsAll(selVars)){		  
		    val l=new SelectionOp(selection.id,join.left,selection.expressions)
		    val r=new SelectionOp(selection.id,join.right,selection.expressions)
		    new InnerJoinOp(staticOptimize(l),staticOptimize(r))
		}
		else selection
	case _=>selection
  }
	

	
  def optimizeGroup(group:GroupOp)= group.subOp match{
    /*case proj:ProjectionOp=>
      val newAggs=group.aggs.map{agg=>
	    val vari=proj.expressions(agg._2.varName).asInstanceOf[VarXpr].varName
	    (agg._1,new AggXpr(agg._2.aggOp,vari))	    
	  }
	  new GroupOp(group.id,group.groupVars,newAggs,proj.subOp)*/
    case _=>group
  }
	


}